File: app/src/main/java/com/baijum/ukufretboard/ui/LearningProgressView.kt
package com.baijum.ukufretboard.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.baijum.ukufretboard.data.LearningStats
import com.baijum.ukufretboard.data.PracticeStats
import com.baijum.ukufretboard.data.PracticeTimerRepository
import com.baijum.ukufretboard.viewmodel.LearningProgressViewModel

/**
 * Learning Progress dashboard showing lesson completion, quiz stats,
 * interval trainer stats, and daily streak.
 */
@Composable
fun LearningProgressView(
    viewModel: LearningProgressViewModel,
    practiceStats: PracticeStats = PracticeStats(),
    modifier: Modifier = Modifier,
) {
    val state by viewModel.state.collectAsState()
    var showResetDialog by remember { mutableStateOf(false) }

    Column(
        modifier = modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 16.dp, vertical = 8.dp),
    ) {
        Text(
            text = "Learning Progress",
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.Bold,
            modifier = Modifier.semantics { heading() },
        )
        Text(
            text = "Track your music theory journey.",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )

        Spacer(modifier = Modifier.height(12.dp))

        // ── Practice Time ──
        if (practiceStats.totalMinutes > 0 || practiceStats.todayMinutes > 0) {
            Card(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(12.dp),
                colors = CardDefaults.cardColors(
                    containerColor = MaterialTheme.colorScheme.tertiaryContainer,
                ),
            ) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Text(
                        text = "Practice Time",
                        style = MaterialTheme.typography.titleSmall,
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.onTertiaryContainer,
                        modifier = Modifier.semantics { heading() },
                    )
                    Spacer(modifier = Modifier.height(8.dp))
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceEvenly,
                    ) {
                        ProgressStat(
                            value = "${practiceStats.todayMinutes}m",
                            label = "Today",
                            color = MaterialTheme.colorScheme.onTertiaryContainer,
                        )
                        ProgressStat(
                            value = practiceStats.totalTimeFormatted,
                            label = "Total",
                            color = MaterialTheme.colorScheme.onTertiaryContainer,
                        )
                        ProgressStat(
                            value = "${practiceStats.totalSessions}",
                            label = "Sessions",
                            color = MaterialTheme.colorScheme.onTertiaryContainer,
                        )
                    }
                    if (practiceStats.dailyGoal > 0) {
                        Spacer(modifier = Modifier.height(8.dp))
                        LinearProgressIndicator(
                            progress = { practiceStats.dailyProgress.coerceAtMost(1f) },
                            modifier = Modifier.fillMaxWidth(),
                            color = MaterialTheme.colorScheme.tertiary,
                            trackColor = MaterialTheme.colorScheme.onTertiaryContainer.copy(alpha = 0.2f),
                        )
                        Text(
                            text = "Daily goal: ${practiceStats.todayMinutes}/${practiceStats.dailyGoal} min",
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onTertiaryContainer,
                            modifier = Modifier.padding(top = 4.dp),
                        )
                    }
                }
            }
            Spacer(modifier = Modifier.height(12.dp))
        }

        // ── Daily Streak ──
        Card(
            modifier = Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(12.dp),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.primaryContainer),
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp),
                horizontalArrangement = Arrangement.SpaceEvenly,
            ) {
                ProgressStat(
                    value = "${state.currentDayStreak}",
                    label = "Day Streak",
                    color = MaterialTheme.colorScheme.onPrimaryContainer,
                )
                ProgressStat(
                    value = "${state.bestDayStreak}",
                    label = "Best Streak",
                    color = MaterialTheme.colorScheme.onPrimaryContainer,
                )
                ProgressStat(
                    value = "${state.lessonCompletionPercent}%",
                    label = "Lessons Done",
                    color = MaterialTheme.colorScheme.onPrimaryContainer,
                )
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        // ── Theory Lessons ──
        SectionHeader("Theory Lessons")
        Card(
            modifier = Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(12.dp),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant),
        ) {
            Column(modifier = Modifier.padding(16.dp)) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                ) {
                    Text(
                        text = "Lessons completed",
                        style = MaterialTheme.typography.bodyMedium,
                    )
                    Text(
                        text = "${state.completedLessons} / ${state.totalLessons}",
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.primary,
                    )
                }
                Spacer(modifier = Modifier.height(8.dp))
                LinearProgressIndicator(
                    progress = {
                        if (state.totalLessons > 0) {
                            state.completedLessons.toFloat() / state.totalLessons
                        } else 0f
                    },
                    modifier = Modifier.fillMaxWidth(),
                    trackColor = MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.3f),
                )
                Spacer(modifier = Modifier.height(8.dp))
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                ) {
                    Text(
                        text = "Quizzes passed",
                        style = MaterialTheme.typography.bodyMedium,
                    )
                    Text(
                        text = "${state.passedLessonQuizzes} / ${state.totalLessons}",
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.primary,
                    )
                }
                Spacer(modifier = Modifier.height(8.dp))
                LinearProgressIndicator(
                    progress = {
                        if (state.totalLessons > 0) {
                            state.passedLessonQuizzes.toFloat() / state.totalLessons
                        } else 0f
                    },
                    modifier = Modifier.fillMaxWidth(),
                    trackColor = MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.3f),
                )
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        // ── Theory Quiz ──
        SectionHeader("Theory Quiz")
        StatsCard(stats = state.quizStatsOverall, label = "Overall")
        Spacer(modifier = Modifier.height(8.dp))
        state.quizStatsByCategory.forEach { (category, stats) ->
            if (stats.total > 0) {
                MiniStatsRow(label = category.label, stats = stats)
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        // ── Interval Trainer ──
        SectionHeader("Interval Trainer")
        StatsCard(stats = state.intervalStatsOverall, label = "Overall")
        Spacer(modifier = Modifier.height(8.dp))
        val levelNames = listOf("Easy", "Medium", "Hard", "Expert")
        state.intervalStatsByLevel.forEach { (level, stats) ->
            if (stats.total > 0) {
                MiniStatsRow(label = levelNames.getOrElse(level - 1) { "Level $level" }, stats = stats)
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        // ── Note Quiz ──
        SectionHeader("Note Quiz")
        StatsCard(stats = state.noteQuizStats, label = "Overall")

        Spacer(modifier = Modifier.height(16.dp))

        // ── Chord Ear Training ──
        SectionHeader("Chord Ear Training")
        StatsCard(stats = state.chordEarStatsOverall, label = "Overall")
        Spacer(modifier = Modifier.height(8.dp))
        state.chordEarStatsByLevel.forEach { (level, stats) ->
            if (stats.total > 0) {
                MiniStatsRow(label = levelNames.getOrElse(level - 1) { "Level $level" }, stats = stats)
            }
        }

        Spacer(modifier = Modifier.height(24.dp))

        // ── Reset ──
        OutlinedButton(
            onClick = { showResetDialog = true },
            modifier = Modifier.fillMaxWidth(),
        ) {
            Text("Reset All Progress")
        }

        Spacer(modifier = Modifier.height(16.dp))
    }

    if (showResetDialog) {
        AlertDialog(
            onDismissRequest = { showResetDialog = false },
            title = { Text("Reset Progress?") },
            text = { Text("This will clear all learning progress, quiz scores, streaks, and lesson completion. This cannot be undone.") },
            confirmButton = {
                TextButton(onClick = {
                    viewModel.clearAllProgress()
                    showResetDialog = false
                }) {
                    Text("Reset", color = MaterialTheme.colorScheme.error)
                }
            },
            dismissButton = {
                TextButton(onClick = { showResetDialog = false }) {
                    Text("Cancel")
                }
            },
        )
    }
}

@Composable
private fun SectionHeader(title: String) {
    Text(
        text = title,
        style = MaterialTheme.typography.titleSmall,
        fontWeight = FontWeight.Bold,
        modifier = Modifier.padding(bottom = 8.dp).semantics { heading() },
    )
}

@Composable
private fun StatsCard(stats: LearningStats, label: String) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant),
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Text(
                text = label,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                fontWeight = FontWeight.Bold,
            )
            Spacer(modifier = Modifier.height(8.dp))
            if (stats.total == 0) {
                Text(
                    text = "No attempts yet.",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            } else {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceEvenly,
                ) {
                    ProgressStat(value = "${stats.correct}/${stats.total}", label = "Score")
                    ProgressStat(value = "${stats.accuracyPercent}%", label = "Accuracy")
                    ProgressStat(value = "${stats.bestStreak}", label = "Best Streak")
                }
            }
        }
    }
}

@Composable
private fun MiniStatsRow(label: String, stats: LearningStats) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 8.dp, vertical = 4.dp),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodySmall,
            modifier = Modifier.weight(1f),
        )
        Text(
            text = "${stats.correct}/${stats.total}",
            style = MaterialTheme.typography.bodySmall,
            fontWeight = FontWeight.Bold,
            modifier = Modifier.padding(horizontal = 12.dp),
        )
        Text(
            text = "${stats.accuracyPercent}%",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.primary,
            fontWeight = FontWeight.Bold,
        )
    }
    HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.2f))
}

@Composable
private fun ProgressStat(
    value: String,
    label: String,
    color: androidx.compose.ui.graphics.Color = MaterialTheme.colorScheme.primary,
) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Text(
            text = value,
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.Bold,
            color = color,
        )
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = color.copy(alpha = 0.7f),
        )
    }
}

File: iosApp/UkuleleCompanion/Views/LearningProgressView.swift
import SwiftUI
import shared

struct LearningProgressView: View {
    @EnvironmentObject var learnVM: LearnViewModel
    @EnvironmentObject var practiceTimerVM: PracticeTimerViewModel
    @State private var showResetConfirmation = false

    var body: some View {
        let _ = learnVM.stateVersion // observe changes

        ScrollView {
            VStack(spacing: 20) {
                // Practice Time
                VStack(alignment: .leading, spacing: 8) {
                    Text("Practice Time")
                        .font(.headline)
                        .accessibilityAddTraits(.isHeader)
                    HStack {
                        VStack(alignment: .leading, spacing: 4) {
                            Text("Total: \(practiceTimerVM.totalTimeFormatted)")
                            Text("Sessions: \(practiceTimerVM.data.totalSessions)")
                            HStack(spacing: 4) {
                                Text("Today: \(practiceTimerVM.todayMinutes)m")
                                Text("/")
                                    .foregroundStyle(.secondary)
                                Text("\(practiceTimerVM.data.dailyGoal)m goal")
                                    .foregroundStyle(.secondary)
                            }
                            Text("Longest: \(practiceTimerVM.data.longestSession)m")
                        }
                        .font(.subheadline)
                        Spacer()
                        VStack {
                            ZStack {
                                Circle()
                                    .stroke(Color(.systemGray4), lineWidth: 6)
                                    .frame(width: 50, height: 50)
                                Circle()
                                    .trim(from: 0, to: min(Double(practiceTimerVM.dailyProgress), 1.0))
                                    .stroke(practiceTimerVM.dailyProgress >= 1.0 ? Color.green : Color.accentColor,
                                            style: StrokeStyle(lineWidth: 6, lineCap: .round))
                                    .frame(width: 50, height: 50)
                                    .rotationEffect(.degrees(-90))
                                Text("\(Int(min(practiceTimerVM.dailyProgress, 1.0) * 100))%")
                                    .font(.caption2.weight(.bold))
                            }
                            Text("Daily Goal")
                                .font(.caption2)
                                .foregroundStyle(.secondary)
                        }
                        .accessibilityCombined(label: "Daily Goal", value: "\(Int(practiceTimerVM.dailyProgress * 100)) percent")
                    }
                }
                .padding()
                .background(Color(.systemGray6))
                .cornerRadius(12)
                .padding(.horizontal)

                // Streak
                HStack(spacing: 24) {
                    statCard(title: "Current Streak", value: "\(learnVM.currentDayStreak())", icon: "flame.fill", color: .orange)
                    statCard(title: "Best Streak", value: "\(learnVM.bestDayStreak())", icon: "trophy.fill", color: .yellow)
                }
                .padding(.horizontal)

                // Lessons
                let totalLessons = (TheoryLessons.shared.ALL as! [TheoryLesson]).count
                let completedLessons = learnVM.completedLessonCount()
                let passedQuizzes = learnVM.passedQuizCount()

                VStack(alignment: .leading, spacing: 8) {
                    Text("Theory Lessons")
                        .font(.headline)
                        .accessibilityAddTraits(.isHeader)
                    HStack {
                        progressCircle(value: Double(completedLessons), total: Double(totalLessons), label: "Completed")
                        Spacer()
                        progressCircle(value: Double(passedQuizzes), total: Double(totalLessons), label: "Quizzes Passed")
                    }
                }
                .padding()
                .background(Color(.systemGray6))
                .cornerRadius(12)
                .padding(.horizontal)

                // Quiz Stats
                statsSection(title: "Theory Quiz", stats: learnVM.quizStats())

                // Interval Stats
                statsSection(title: "Interval Trainer", stats: learnVM.intervalStats())

                // Note Quiz
                statsSection(title: "Note Quiz", stats: learnVM.noteQuizStats())

                // Chord Ear Training
                statsSection(title: "Chord Ear Training", stats: learnVM.chordEarStats())

                // Scale Practice
                statsSection(title: "Scale Practice", stats: learnVM.scalePracticeStats())

                Button(role: .destructive) {
                    showResetConfirmation = true
                } label: {
                    Label("Reset All Progress", systemImage: "arrow.counterclockwise")
                        .frame(maxWidth: .infinity)
                }
                .buttonStyle(.bordered)
                .padding(.horizontal)
                .padding(.top, 8)
            }
            .padding(.vertical)
        }
        .navigationTitle("Learning Progress")
        .confirmationDialog(
            "Reset All Progress?",
            isPresented: $showResetConfirmation,
            titleVisibility: .visible
        ) {
            Button("Reset", role: .destructive) {
                learnVM.clearAllProgress()
            }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("This will permanently delete all your learning progress, streaks, and quiz statistics. This action cannot be undone.")
        }
    }

    private func statCard(title: String, value: String, icon: String, color: Color) -> some View {
        VStack(spacing: 8) {
            Image(systemName: icon)
                .font(.title)
                .foregroundStyle(color)
            Text(value)
                .font(.title.weight(.bold))
            Text(title)
                .font(.caption)
                .foregroundStyle(.secondary)
        }
        .frame(maxWidth: .infinity)
        .padding()
        .background(Color(.systemGray6))
        .cornerRadius(12)
        .accessibilityCombined(label: title, value: value)
    }

    private func progressCircle(value: Double, total: Double, label: String) -> some View {
        VStack(spacing: 4) {
            ZStack {
                Circle()
                    .stroke(Color(.systemGray4), lineWidth: 6)
                    .frame(width: 60, height: 60)
                Circle()
                    .trim(from: 0, to: total > 0 ? value / total : 0)
                    .stroke(Color.accentColor, style: StrokeStyle(lineWidth: 6, lineCap: .round))
                    .frame(width: 60, height: 60)
                    .rotationEffect(.degrees(-90))
                Text("\(Int(value))/\(Int(total))")
                    .font(.caption2.weight(.medium))
            }
            Text(label)
                .font(.caption2)
                .foregroundStyle(.secondary)
        }
        .accessibilityCombined(label: label, value: "\(Int(value)) of \(Int(total))")
    }

    private func statsSection(title: String, stats: LearningStats) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(title)
                .font(.headline)
                .accessibilityAddTraits(.isHeader)
            HStack {
                VStack(alignment: .leading) {
                    Text("Total: \(stats.total)")
                    Text("Correct: \(stats.correct)")
                    Text("Best Streak: \(stats.bestStreak)")
                }
                .font(.subheadline)
                Spacer()
                if stats.total > 0 {
                    VStack {
                        ZStack {
                            Circle()
                                .stroke(Color(.systemGray4), lineWidth: 6)
                                .frame(width: 50, height: 50)
                            Circle()
                                .trim(from: 0, to: Double(stats.accuracyPercent) / 100)
                                .stroke(accuracyColor(stats.accuracyPercent), style: StrokeStyle(lineWidth: 6, lineCap: .round))
                                .frame(width: 50, height: 50)
                                .rotationEffect(.degrees(-90))
                            Text("\(stats.accuracyPercent)%")
                                .font(.caption2.weight(.bold))
                        }
                        Text("Accuracy")
                            .font(.caption2)
                            .foregroundStyle(.secondary)
                    }
                    .accessibilityCombined(label: "Accuracy", value: "\(stats.accuracyPercent) percent")
                }
            }
        }
        .padding()
        .background(Color(.systemGray6))
        .cornerRadius(12)
        .padding(.horizontal)
    }

    private func accuracyColor(_ percent: Int) -> Color {
        if percent >= 80 { return .green }
        if percent >= 60 { return .yellow }
        return .orange
    }
}
